### PR TITLE
Introduce Helidon Pico

### PR DESCRIPTION
**Pull request by trentjeff**
This PR introduces Pico, specifically the API and Types modules.

### CODE REVIEW

File: .gitignore
Comment (tomas-langer) [inline .gitignore:26]: This file should be generated in target directory, which is already excluded, why add it again?

File: pico/.gitignore
Comment (tomas-langer) [inline pico/.gitignore:1]: This file does not belong here. There is only one git ignore for the repo, in the root

File: pico/api/src/main/java/module-info.java
Comment (tomas-langer) [inline pico/api/src/main/java/module-info.java:18]: This should explain what Pico is and where to start

File: pico/pom.xml
Comment (tomas-langer) [inline pico/pom.xml:44]: No commented out lines should be in our production code

File: pico/pom.xml
Comment (tomas-langer) [inline pico/pom.xml:83]: This only belongs to root pom file of Helidon, we do not have any other developer attributions (git is the source of truth)

File: pico/types/pom.xml
Comment (tomas-langer) [inline pico/types/pom.xml:31]: Please use the convention of other modules - should be artifact id in english (in this case "Helidon Pico Types".
Add additional details to description if you want to

**Review by tomas-langer — CHANGES_REQUESTED** [review]
This seems not to contain any code. This should not work with our pipelines (Javadoc should fail on this). Also hard to approve module names when we do not know what is the content.

File: .gitignore
Comment (trentjeff) [reply .gitignore:26]: I added it so that I can hide the pico modules locally that are not yet ready to merge down.

File: pico/.gitignore
Comment (trentjeff) [reply pico/.gitignore:1]: As I mentioned previously, I want to hide files not ready to merge down.

File: pico/api/src/main/java/io/helidon/pico/api/ExternalContracts.java
Comment (tomas-langer) [inline pico/api/src/main/java/io/helidon/pico/api/ExternalContracts.java:31]: We should not need runtime retention, as we plan to do everything at compile time (this is valid for all the annotations).

File: pico/api/src/main/java/module-info.java
Comment (trentjeff) [reply pico/api/src/main/java/module-info.java:18]: The readme will explain that - which will be in a future pr.

File: pico/types/src/main/java/io/helidon/pico/types/AnnotationAndValue.java
Comment (tomas-langer) [inline pico/types/src/main/java/io/helidon/pico/types/AnnotationAndValue.java:39]: Naming of methods is not according to Helidon code style. We use "record" style (valid for all methods)

**Review by tomas-langer — CHANGES_REQUESTED** [review]

File: .gitignore
Comment (tomas-langer) [reply .gitignore:26]: I was talking about `javac.*.args` - that exclusion should not be needed. Which files match it that you need to add it to exclusion?

File: pico/.gitignore
Comment (tomas-langer) [reply pico/.gitignore:1]: Please remove this file. If you "un-add" your files from git, they will not be added to your commits and pushes.

File: pico/api/src/main/java/module-info.java
Comment (tomas-langer) [reply pico/api/src/main/java/module-info.java:18]: `module-info.java` and `package-info.java` are the only things to be seen when you access Javadoc. So it is good to add some basic documentation to them (at least where to start).

File: pico/api/src/main/java/io/helidon/pico/api/ExternalContracts.java
Comment (trentjeff) [reply pico/api/src/main/java/io/helidon/pico/api/ExternalContracts.java:31]: I've already advised the developer to not include this module into compile scope; my thinking is to leave it here for anyone that still wants to use reflection, etc.

File: pico/pom.xml
Comment (tomas-langer) [inline pico/pom.xml:43]: not used

File: pico/types/pom.xml
Comment (tomas-langer) [inline pico/types/pom.xml:35]: these two dependencies are not used

**Review by tomas-langer — CHANGES_REQUESTED** [review]

File: pico/api/src/main/java/io/helidon/pico/api/ExternalContracts.java
Comment (tomas-langer) [reply pico/api/src/main/java/io/helidon/pico/api/ExternalContracts.java:31]: I think we should not do that - it will make the class file bigger and we want to process it compile time always.

File: pico/types/src/main/java/io/helidon/pico/types/AnnotationAndValue.java
Comment (trentjeff) [reply pico/types/src/main/java/io/helidon/pico/types/AnnotationAndValue.java:70]: For optional string attributes on annotations, most people as a convention will use 'default ""'. That does not technically count as having a value.  There are also many places in code coming that will need the "not null and not blank" API like this.  We can think about renaming it, but taking it out is not an option.

File: pico/types/src/main/java/io/helidon/pico/types/AnnotationAndValue.java
Comment (trentjeff) [reply pico/types/src/main/java/io/helidon/pico/types/AnnotationAndValue.java:62]: will make the javadoc clearer

File: pico/types/src/main/java/io/helidon/pico/types/AnnotationAndValue.java
Comment (tomas-langer) [reply pico/types/src/main/java/io/helidon/pico/types/AnnotationAndValue.java:71]: If you look at our own annotations, we do use empty strings a lot of time (when feasible), but sometimes we had to implement a custom default value, to allow usage of empty string as a valid value.
So yes, in a lot of cases empty string means not having a value. But is some cases it does not.

File: pico/types/src/main/java/io/helidon/pico/types/TypeName.java
Comment (tomas-langer) [reply pico/types/src/main/java/io/helidon/pico/types/TypeName.java:51]: Such an annotation would only make sense if you decide to send `TypeName` through Jackson. As that is not the case, the annotations are useless for target teams. 
This type should only be used by Helidon team (or team extending Pico).

File: .gitignore
Comment (trentjeff) [reply .gitignore]: took that out and pushed.

File: pico/.gitignore
Comment (trentjeff) [reply pico/.gitignore]: removed and pushed.

File: pico/api/src/main/java/module-info.java
Comment (trentjeff) [reply pico/api/src/main/java/module-info.java:18]: pushed updates to package-info; please review again.

File: pico/types/src/main/java/io/helidon/pico/types/AnnotationAndValue.java
Comment (trentjeff) [reply pico/types/src/main/java/io/helidon/pico/types/AnnotationAndValue.java:39]: I fixed the javadoc.

File: pico/types/src/main/java/io/helidon/pico/types/TypeName.java
Comment (ljnelson) [reply pico/types/src/main/java/io/helidon/pico/types/TypeName.java:51]: (Neutral on all this; just here to note the existence of [Jackson mixins](https://medium.com/@shankar.ganesh.1234/jackson-mixin-a-simple-guide-to-a-powerful-feature-d984341dc9e2), which is the normal way someone can "put" a Jackson annotation on a class that does not or cannot have Jackson annotations on it.)

File: pico/types/src/main/java/io/helidon/pico/types/AnnotationAndValue.java
Comment (trentjeff) [reply pico/types/src/main/java/io/helidon/pico/types/AnnotationAndValue.java:71]: First, I need a method that will do this check in its implementation somewhere in the API. If not here then I'll need to pick a new spot for that impl. This avoids having to write "val != null && !val.isBlank()" in many, many places that will soon be coming.  I can conceptually use "empty" instead of "blank" with no adverse effects, but I thought blank might be better to use in this case. This is point 1.

Second, are you suggesting we need to distinguish *at compile time* a need to differentiate the default value of "" vs intended/ascribed value of "" by the developer? If so then that simply doesn't exist, nor do I believe have any need for it since the processing will behave the same regardless of how it was originated using defaults or via an override.  If it's blank (or empty) it doesn't have a value - plain and simple.

So the realistic options I see are:
0) Use javadoc to clearly describe how "" is considered to not have any value from our opinionated PoV on this class at leasy.
1) use "empty" instead of "blank" for hasValue()
2) use a better description: hasValue() -> hasNonBlankValue() and/or hasNonEmptyValue()

File: pico/types/src/main/java/io/helidon/pico/types/TypeName.java
Comment (trentjeff) [reply pico/types/src/main/java/io/helidon/pico/types/TypeName.java:51]: Chances are better than not that consumers of Helidon are using Jackson and Lombok.

And in such cases then we are implicitly hurting them by not including JsonIgnore.  I therefore would maintain my point that it doesn't harm anyone (internally or externally) by us having a provided/static module dependency here, but it can in fact hurt our customers that may be using it in these scenarios where they are using Jackson, Lombok, etc.

See https://medium.com/@mrkrijgsman/flexible-immutability-a475346d9518

Comment (m0mus) [reply pico/types/src/main/java/io/helidon/pico/types/TypeName.java:51]: What about defining this field as `transient`?

File: pico/types/src/main/java/io/helidon/pico/types/AnnotationAndValue.java
Comment (spericas) [reply pico/types/src/main/java/io/helidon/pico/types/AnnotationAndValue.java:39]: Right, these should be called `typeName()` and `value()` to be consistent with Helidon conventions.

File: pico/types/src/main/java/io/helidon/pico/types/TypeName.java
Comment (spericas) [reply pico/types/src/main/java/io/helidon/pico/types/TypeName.java:51]: I agree, it's odd to use Jackson annotations here. But perhaps the PR is too narrow focused to understand the bigger picture.

File: pico/types/src/main/java/io/helidon/pico/types/AnnotationAndValue.java
Comment (trentjeff) [reply pico/types/src/main/java/io/helidon/pico/types/AnnotationAndValue.java]: I still think this is the wrong decision, just for the record. These APIs were meant to mirror TypeName -> https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/Type.html and AnnotationAndValue -> https://docs.oracle.com/javase/8/docs/api/javax/lang/model/element/AnnotationValue.html

But it seems like consistency to the rest of Helidon for the API contracts is paramount, so I ended up pushing the suggested changes.

File: pico/types/src/main/java/io/helidon/pico/types/TypeName.java
Comment (trentjeff) [reply pico/types/src/main/java/io/helidon/pico/types/TypeName.java]: i took it out of the contracts for this PR, but we will see these again when we get to the impl modules :-)

File: pico/api/src/main/java/io/helidon/pico/api/Contract.java
Comment (trentjeff) [reply pico/api/src/main/java/io/helidon/pico/api/Contract.java:26]: i thought it was 130?  we have a checksyle failure for 130.

Comment (spericas) [reply pico/api/src/main/java/io/helidon/pico/api/Contract.java:26]: Yes, max is 130, but I thought 120 was recommended. Interestingly, `RunLevel.java` seems to be over but is passing the check.

Comment (trentjeff) [reply pico/api/src/main/java/io/helidon/pico/api/Contract.java:26]: i will address if i have to make any more changes, otherwise will leave as-is thanks!

Comment (tomas-langer) [reply pico/api/src/main/java/io/helidon/pico/api/Contract.java:26]: Let's leave this to the checkstyle setup. 
We allo longer lines if the line is a link (and a few other)

**Review by tomas-langer — CHANGES_REQUESTED** [review]

File: pico/types/src/main/java/io/helidon/pico/types/AnnotationAndValue.java
Comment (trentjeff) [reply pico/types/src/main/java/io/helidon/pico/types/AnnotationAndValue.java]: renamed to value() and pushed.

File: pico/types/src/main/java/io/helidon/pico/types/TypeName.java
Comment (trentjeff) [reply pico/types/src/main/java/io/helidon/pico/types/TypeName.java]: renamed to name() and pushed.

**Review by tomas-langer — APPROVED** [review]
LGTM

**Review by spericas — APPROVED** [review]

**Pull request merged**
